Migrate `codegen_ssa` to diagnostics structs - [Part 1]

JhonnyBillM opened the pull request:
Initial migration of `codegen_ssa`. Going to split this crate migration in at least two PRs in order to avoid a huge PR and to quick off some questions around:

1. Translating messages from "external" crates.
2. Interfacing with OS messages.
3. Adding UI tests while migrating diagnostics.

_See comments below._

JhonnyBillM committed:
ADD - migrate lib.def write fatal error

This diagnostic has no UI test 🤔 Should we add some? If so, how?

JhonnyBillM committed:
UPDATE - LibDefWriteFailure to accept type instead of formatted string

This follows team’s suggestions in this thread https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n/topic/.23100717.20diag.20translation/near/295305249

JhonnyBillM committed:
UPDATE - migrate write.rs to new diagnostics infra

JhonnyBillM committed:
UPDATE - codege-ssa errors to new Diagnostic macro name

JhonnyBillM committed:
UPDATE - resolve fixme and emit errors via Handler

JhonnyBillM committed:
DELETE - unused error after PR# 100101 was merged

JhonnyBillM committed:
ADD - initial port of link.rs

rust-highfive commented:
r? @eholk

(rust-highfive has picked a reviewer for you, use r? to override)

JhonnyBillM commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -0,0 +1,197 @@
+}
+pub struct ThorinErrorWrapper(pub thorin::Error);
+
+// FIXME: How should we support translations for external crate errors?
+impl IntoDiagnosticArg for ThorinErrorWrapper {
+    fn into_diagnostic_arg(self) -> DiagnosticArgValue<'static> {
+        DiagnosticArgValue::Str(Cow::Owned(format!("{:?}", self.0)))
+    }
+}
```
This `thorin::Error` diagnostic is currently untranslatable. [thorin::Error already implements Display](https://github.com/rust-lang/thorin/blob/c454f85084a882c3905d838b0178b5e2953fe08e/thorin/src/error.rs#L167), which brings the question:

### How do we want to support translatable diagnostics when using an external crate ?

I think the correct path could be for Errors to be able to provide their own translated message, now, this probably involves additions to the standard library and a long-term process for this practice to be adopted in the ecosystem.

Another solution could be for `rustc` to handle this internally for the crate it uses, for example, we could match over the error cases and provide our own keys and translations, but this is sub-optional to say the least. 
`rustc` should not be maintaining translations for other crates, and at the end this is something that is going to impact the ecosystem once the compiler creates the expectation of having i18n interfaces.

What do you folks think?

davidtwco replied:
`thorin` is a crate that I wrote and that lives in `rust-lang/thorin` so we can make whatever changes we need to it - in general, I'm hoping that we can pattern match on error types and just map those to whatever Fluent messages we want.

JhonnyBillM replied:
Sounds ok, but we risk getting diagnostics out of sync between the two repositories

JhonnyBillM replied:
Do we want to include all of `thorin::Error` keys `codegen_ssa.ftl` file, or do you think it would be better to have `thorin.ftl` and a separate `thorin_error.rs` file inside `codegen_ssa` ?

I prefer the latter, but I think our "rule" so far has been one `.ftl` file per crate.

eholk replied:
This is probably a naive question, but is it possible to put the `.ftl` file in the thorin crate?

davidtwco replied:
I guess we could add a constant to Thorin with the ftl content but I don't think we'd gain too much from doing that, because we'd then want to run it through the `fluent_messages!` macro to generate our typed identifiers and stuff like that (which would require some changes). For now, we can just keep them in `codegen_ssa.ftl`, `thorin` is unlikely to many new errors or have its messages change. There's a one-off task in #100717 for keeping each ftl file with its crate rather than in `rustc_error_messages` - whatever we do for that would likely resolve this too.

JhonnyBillM replied:
Addressed [here](https://github.com/rust-lang/rust/pull/102612/commits/f69f1b0fc8f8d730798b9633f0f1114d96259bf4)

JhonnyBillM commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -0,0 +1,197 @@
+
+        diag.note(format!("{:?}", self.command)).note(self.escaped_output);
+
+        // Trying to match an error from OS linkers
+        // which by now we have no way to translate.
+        if self.escaped_output.contains("undefined reference to") {
+            diag.note(fluent::codegen_ssa::extern_funcs_not_found)
+                .note(fluent::codegen_ssa::specify_libraries_to_link)
+                .note(fluent::codegen_ssa::use_cargo_directive);
+        }
```
We know that `undefined reference to` is going to be in English because it is coming from the linker, so this will continue to work even in a different locale, but ideally we should be moving everything to not assume English as the default language.
 
Would you suggest handling this in another way?

JhonnyBillM commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -0,0 +1,197 @@
+
+        diag.note(format!("{:?}", self.command)).note(self.escaped_output);
+
+        // Trying to match an error from OS linkers
+        // which by now we have no way to translate.
+        if self.escaped_output.contains("undefined reference to") {
```
We know that `undefined reference to` is going to be in English because it is coming from the linker, so this will continue to work even in a different locale, but ideally we should be moving everything to not assume English as the default language.
 
Would you suggest handling this in another way?

Nilstrieb replied:
This won't work. On my Windows 11, linker errors are localized already, and ```rust
extern {
    fn aaa();
}

fn main() {
    unsafe { aaa() }
}
```
gives me the following error (with broken encoding, as is tradition with computers)
```
error LNK2019: Verweis auf nicht aufgel\x94stes externes Symbol \"aaa\" in Funktion \"_ZN1x4main17h8b46ac714b2db20fE\".\r\nx.exe : fatal error LNK1120: 1 nicht aufgel\x94ste Externe\r\n
```

Nilstrieb replied:
(this is extra funny because my current OS language is English, but looks like link.exe was originally configured in German and didn't bother changing its language when I changed my OS language)

Nilstrieb replied:
Also, `mold`  (and according to a quick internet search, `lld` as well) will error with `mold: error: undefined symbol: aaa` which does not contain `undefined reference to`

davidtwco replied:
I think we can leave this logic as-is right now, it's imperfect but I don't think we can do much better and it's tangential to diagnostic migration.

JhonnyBillM commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -87,13 +88,13 @@ pub fn get_linker<'a>(
                 } else {
-                    warn!("arch is not supported");
+                    sess.emit_warning(errors::UnsupportedArch);
                 }
             } else {
-                warn!("MSVC root path lib location not found");
+                sess.emit_warning(errors::MsvcPathNotFound);
             }
         } else {
-            warn!("link.exe not found");
+            sess.emit_warning(errors::LinkExeNotFound);
```
Should these migrations be reversed ?

davidtwco replied:
Yes, these are logging statements, not diagnostics.

JhonnyBillM replied:
Reverted [here](https://github.com/rust-lang/rust/pull/102612/commits/8d94cda8da8e3350f972d5c7a06dae6dc90b4fdd)

davidtwco commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -385,8 +380,11 @@ fn link_rlib<'a>(
-            sess.fatal(&format!("failed to add native library {}: {}", output_path.display(), e));
+        ab.add_archive(&output_path, Box::new(|_| false)).unwrap_or_else(|error| {
+            sess.emit_fatal(errors::AddNativeLibrary {
+                library_path: &output_path.display().to_string(),
```
I'm pretty sure we support `Path` and `PathBuf` fields in the structs.

JhonnyBillM replied:
Yes, but this error is being used in two places, and one of them uses `.to_string_lossy()` instead of PathBuf's Display implementation.

```rust
                sess.emit_fatal(errors::AddNativeLibrary {
                    library_path: &location.to_string_lossy(),
                    error,
                });
```

In this case, would you prefer to have two errors instead, or is it fine as is ?

davidtwco replied:
I would be surprised if the difference is important, you can probably just use `PathBuf`.

JhonnyBillM replied:
Updated [here](https://github.com/rust-lang/rust/pull/102612/commits/954a62ff46cb8d6768e28a467e4393a780886576)

davidtwco commented on compiler/rustc_error_messages/src/lib.rs:
```diff
@@ -62,6 +62,7 @@ fluent_messages! {
     typeck => "../locales/en-US/typeck.ftl",
     mir_dataflow => "../locales/en-US/mir_dataflow.ftl",
     symbol_mangling => "../locales/en-US/symbol_mangling.ftl",
+    codegen_ssa => "../locales/en-US/codegen_ssa.ftl",
```
Can you sort this list and add the new line at the appropriate point?

JhonnyBillM replied:
Updated [here](https://github.com/rust-lang/rust/pull/102612/commits/8d94cda8da8e3350f972d5c7a06dae6dc90b4fdd)

davidtwco reviewed (changes requested):
This is a great start, thanks!

JhonnyBillM commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -0,0 +1,197 @@
+    MissingFormat,
+    OnlyRmetaFound { crate_name: Symbol },
+    NotFound { crate_name: Symbol },
+}
```
Oh wow, didn't know that 👏 
Addressed [here](https://github.com/rust-lang/rust/pull/102612/commits/8d94cda8da8e3350f972d5c7a06dae6dc90b4fdd)

Thanks

eholk commented:
Looks good to me, but I'll hand it off to @davidtwco to make sure he feels his comments are adequately addressed.

@bors r? @davidtwco

bors commented:
:umbrella: The latest upstream changes (presumably #102395) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

JhonnyBillM committed:
ADD - implement IntoDiagnostic for thorin::Error wrapper

JhonnyBillM committed:
ADD - codegen_ssa initial diags translations machinery

ADD - migrate MissingNativeStaticLibrary fatal error

JhonnyBillM committed:
Address PR comments

- UPDATE - revert migration of logs

- UPDATE - use derive on LinkRlibError enum

- [Gardening] UPDATE - alphabetically sort fluent_messages

- UPDATE - use PathBuf and unify both AddNativeLibrary to use Display (which is what PathBuf uses when conforming to IntoDiagnosticArg)

- UPDATE - fluent messages sort after rebase

JhonnyBillM commented on compiler/rustc_codegen_ssa/src/errors.rs:
```diff
@@ -0,0 +1,353 @@
+                diag.set_arg("error", format!("{e}"));
+                diag
+            }
+            _ => unimplemented!("Untranslated thorin error"),
```
Is there any specific way to handle `non_exhaustive` enums in `rustc`?

davidtwco replied:
I think this is the best we can do.

davidtwco commented on compiler/rustc_error_messages/locales/en-US/codegen_ssa.ftl:
```diff
@@ -0,0 +1,119 @@
+
+codegen_ssa_use_cargo_directive = use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
+
+codegen_ssa_thorin_read_input_failure = Failed to read input file
```
Can you lower-case the thorin errors? We might as well follow rustc convention if we're writing them here.

JhonnyBillM replied:
Updated [here](https://github.com/rust-lang/rust/pull/102612/commits/41ab94cb00f0e08590660260d20d4ba534042a88)

davidtwco commented:
@bors r+

bors commented:
:umbrella: The latest upstream changes (presumably #101988) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

JhonnyBillM committed:
UPDATE - migrate linker.rs to new diagnostics infra

JhonnyBillM commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 13d4f27c829a332f50a2109647f3b16bce119b8d has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 13d4f27c829a332f50a2109647f3b16bce119b8d -->
<!-- homu: {"type":"Approved","sha":"13d4f27c829a332f50a2109647f3b16bce119b8d","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
